packages/protoplugin-example/protoc-gen-twirp-es/src/typescript.ts

```diff
@@ -0,0 +1,110 @@
+    f.print("    async request<T extends Message<T>>(");
+    f.print("        service: string,");
+    f.print("        method: string,");
+    f.print("        contentType: string,");
+    f.print("        data: T");
+    f.print("    ) {");
+    f.print("        const headers = new Headers(this.options.headers ?? []);");
+    f.print("        headers.set('content-type', contentType);");
+    f.print("        const response = await fetch(");
+    f.print("            `${this.options.baseUrl}/${service}/${method}`,");
+    f.print("            {");
+    f.print("                ...this.options,");
+    f.print("                method: 'POST',");
+    f.print("                headers,");
+    f.print("                body: data.toJsonString(),");
+    f.print("            }");
+    f.print("        );");
+    f.print("        if (response.status === 200) {");
+    f.print("            if (contentType === 'application/json') {");
+    f.print("                return await response.json();");
+    f.print("            }");
+    f.print("            return new Uint8Array(await response.arrayBuffer());");
+    f.print("        }");
+    f.print("        throw Error(await response.json());");
+    f.print("    }");
+    f.print("}");
+    f.print();
+    for (const service of file.services) {
+      const localServiceName = localName(service);
+      f.print("export function create", localServiceName, "Client(opts: TransportOptions): ",localServiceName, " {");
+      f.print("    return new ", localServiceName, "Client(opts);");
+      f.print("}");
+      f.print();
+      f.print("export interface ", localServiceName, " {");
+      for (const method of service.methods) {
+        f.print(makeJsDoc(method, "    "));
+        f.print("    ", method.name, "(request: ", method.input, "): Promise<", method.output, ">;");
+      }
+      f.print("}");
+      f.print();
+      f.print("export class ", localServiceName, "Client extends TwirpClient {");
+      f.print("    constructor(opts: TransportOptions) {");
+      f.print("        super(opts);");
+      f.print("    }");
+      f.print();
+      for (const method of service.methods) {
+        f.print("    async ", method.name, "(request: ", method.input, "): Promise<", method.output, "> {");
+        if (method.methodKind === MethodKind.Unary) {
+            f.print("        const promise = this.request(");
+            f.print("            ", literalString(service.typeName), ", ");
+            f.print("            ", literalString(method.name), ",");
+            f.print('            "application/json",');
+            f.print("            request");
+            f.print("        );");
+            f.print("        return promise.then(async (data) =>");
+            f.print("             ", method.output, ".fromJson(data as JsonValue)");
+            f.print("        );");
+        } else {
+            f.print("        throw new Error('", MethodKind[method.methodKind]," is not supported');");
```

Comment:
Twirp really only specifies behavior for unary RPCs. I think it's best to skip others.